---
title: "barcoding_analysis"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: fill
editor_options: 
  chunk_output_type: inline
---



```{r setup, include=FALSE}

# clear global enviroment
rm(list = ls())

# initiate timer
start.time <- proc.time()

# List of packages to load
packages2load <- c("Seurat", "GSVA", 
                   "plyr", "dplyr", "tidyr", "reshape2", "preprocessCore", "stringr", "RColorBrewer", "ggplot2", "gridExtra", 
                   "DT", "flexdashboard", "ggpmisc", "ggExtra", "grid", "ggrepel", "future", "scMiko", "survival", "survminer", "topGO",
                   "AnnotationDbi", "org.Mm.eg.db", "org.Hs.eg.db", "ddpcr", "tm", "homologene", "plotly", "parallel", "doParallel", "foreach", "quantreg")

# load packages
invisible(lapply(packages2load, library, character.only = TRUE, quietly = T))


```

```{r README}

# LOAD R ENVIROMENT FROM PART 2 PRIOR TO RUNNING THIS SCRIPT

```


```{r TCGA analysis}

# load TCGA data (local) #######################################################
if (!exists("df.tcga")) df.tcga <- readRDS("C:/Users/Owner/Dropbox/PDF Projects - JM/Data/tcga/rna_expression_tcga_200420.rds")
if (!exists("tcga.meta")) tcga.meta <- readRDS("C:/Users/Owner/Dropbox/PDF Projects - JM/Data/tcga/patient_metaData_tcga_200420.rds")

# filter gbm samples
which.patients <- tcga.meta[["name"]] %in% "Glioblastoma Multiforme"
gbm.meta <- tcga.meta[which.patients]
gbm.mat <- t(df.tcga[ which.patients, ])
gbm.mat <- apply(gbm.mat, 2, function(x) (log1p(x) - mean(log1p(x))) / sd(log1p(x)))

# meta features of interest 
gbm.meta.sub <- data.frame(
  dtd = tcga.meta[["days_to_death"]],
  dtf = tcga.meta[["days_to_last_follow_up"]],
  tumor.grade = tcga.meta[["paper_tumor_grade"]],
  histology = tcga.meta[["paper_Histology"]],
  age = tcga.meta[["age_at_diagnosis"]],
  idh.subtype = tcga.meta[["paper_IDH.status"]],
  mgmt.promoter = tcga.meta[["paper_MGMT.promoter.status"]],
  ch7.10 = tcga.meta[["paper_Chr.7.gain.Chr.10.loss"]],
  ch19.20 = tcga.meta[["paper_Chr.19.20.co.gain"]],
  o.subtype = tcga.meta[["paper_Original.Subtype"]],
  t.subtype = tcga.meta[["paper_Transcriptome.Subtype"]],
  glioma.cluster = tcga.meta[["paper_Pan.Glioma.RNA.Expression.Cluster"]],
  idh.cluster = tcga.meta[["paper_IDH.specific.RNA.Expression.Cluster"]],
  rf.class = tcga.meta[["paper_Random.Forest.Sturm.Cluster"]],
  alive = tcga.meta[["vital_status"]]
)

gbm.meta.sub$time <- gbm.meta.sub$dtf
gbm.meta.sub$time[!is.na(gbm.meta.sub$dtd)] <-gbm.meta.sub$dtd[!is.na(gbm.meta.sub$dtd)]

gbm.meta.sub <- gbm.meta.sub[which.patients, ]

gbm.meta.sub$alive[is.na(gbm.meta.sub$dtd) & !is.na(gbm.meta.sub$dtf)] <- "Alive"

so.tcga <- CreateSeuratObject(counts = gbm.mat)

so.tcga <- ScaleData(so.tcga,  do.scale = F, do.center = F)
so.tcga <- FindVariableFeatures(so.tcga)
so.tcga <- RunPCA(so.tcga)
so.tcga <- FindNeighbors(so.tcga)
so.tcga <- FindClusters(so.tcga)
so.tcga <- RunUMAP(so.tcga, dims = 1:30)
cluster.UMAP(so.tcga)
# so.tcga@meta.data[["nFeature_RNA"]] <- gbm.meta.sub

# tcga scoring helper function #################################################
TCGAscores <- function(gbm.mat,so.sig.set = NULL, gbm.meta.sub, deg.gene.sets = NULL){
  
  if (!is.null(gbm.meta.sub) & is.null(deg.gene.sets)){
    deg.gene.sets <- list(
      setA = so.sig.set$genes[so.sig.set$direction == 1],
      setB = so.sig.set$genes[so.sig.set$direction == -1]
    )
  } 
  
  tcga.gsva.scores <- GSVA::gsva(as.matrix(gbm.mat), deg.gene.sets)
  df.tcga.gsva.score <- data.frame(y1 = tcga.gsva.scores[1,], y2 = tcga.gsva.scores[2,], sample = colnames(tcga.gsva.scores))
  df.tcga.gsva.score$line <- stringr::str_extract(df.tcga.gsva.score$sample, "[A-Za-z0-9]*")
  
  df.tcga.gsva.score.final <- bind_cols(df.tcga.gsva.score, gbm.meta.sub)
  
  return(df.tcga.gsva.score.final)
}

```


```{r Differential pathway analysis - single cell data, fig.height=5, fig.width=6}

# get GSVA pathway scores ######################################################
gsva.mat <- so.query@misc[["gsva"]][["gsva.data"]]
gsva.mat <- gsva.mat[ ,colnames(gsva.mat) %in% colnames(so.query)]

# load into seurat object
so.gsva <- CreateSeuratObject(
  counts = gsva.mat,
  project = "module.score.DEG",
  assay = "RNA",
  min.cells = 0,
  min.features = 0,
  names.field = 1,
  names.delim = "_",
  meta.data = NULL
)
so.gsva <- FindVariableFeatures(so.gsva, selection.method = "mvp")

# get original seurat clusters
so.gsva@meta.data[["seurat_clusters"]] <- so.query@meta.data[["seurat_clusters"]]
so.gsva@meta.data[["Barcode"]] <- so.query@meta.data[["Barcode"]]
Idents(so.gsva) <- so.gsva@meta.data[["Barcode"]]

# run DEG analysis on pathway scores ###########################################
deg.gsva <- FindMarkers(so.gsva,
             ident.1 = group.1.name,
             ident.2 = group.2.name,
             test.use = "MAST",
             logfc.threshold = 0.1)

deg.gsva$term <- rownames(deg.gsva)
deg.gsva.sig <- deg.gsva[deg.gsva$p_val_adj < 0.1, ]

deg.gsva %>%
  ggplot(aes(x = avg_log2FC, y = -log10(p_val), label =  term)) + 
  geom_point()

deg.gsva.sig$term.GO <- stringr::str_extract(deg.gsva.sig$term, "GO:[0-9]*")
deg.gsva.sig$term.clean <- deg.gsva.sig$term
deg.gsva.sig$term.clean <- firstup(stringr::str_remove(deg.gsva.sig$term, "%[A-Z]*%GO:[0-9]*"))

deg.gsva.sig$term.clean2 <- stringr::str_trunc(stringr::str_extract(deg.gsva.sig$term, "[0-9a-zA-Z- ]*"), 30)
deg.gsva.sig.go <- deg.gsva.sig %>% dplyr::filter(!is.na(term.GO))
deg.gsva.sig.go$term.clean3 <- paste0(deg.gsva.sig.go$term.clean, "_",deg.gsva.sig.go$term.GO)
deg.gsva.sig.top <- bind_rows((deg.gsva.sig.go %>% top_n(5, avg_log2FC)), (deg.gsva.sig.go %>% top_n(5, -avg_log2FC)))

# generate ranked differential pathway plot ####################################
plt.gsva.pathway <- deg.gsva.sig.top %>%
  filter(!is.na(term.GO)) %>%
   ggplot(aes(x = reorder(stringr::str_wrap((term.clean3), width = 50), avg_log2FC), y = avg_log2FC, fill = avg_log2FC)) + 
  geom_segment(aes(yend = 0, xend = reorder(stringr::str_wrap((term.clean3), width = 50), avg_log2FC))) + 
   geom_point(aes(size = -log10(p_val_adj)), pch = 21) + 
   coord_flip() + 
   xlab("Pathways") + ylab("logFC (MBT06-BT799)") + 
   labs(title = "Differential Pathways", subtitle = ("BT799 vs. MBT06 (controls)"), fill = "logFC", size = "-log10(FDR)") + 
   geom_hline(yintercept = 0, linetype = "dashed") + 
  theme_miko(legend = T) + 
  scale_fill_distiller(palette = "RdBu")

plt.gsva.pathway
# savePDF("final_differerential_pathways_040321.pdf", plt.gsva.pathway, fig.height=5, fig.width=6)


```



```{r module pathways gsva BULK RNA}


do.bulk <- F

if (do.bulk){
  # bulk RNA gsva analysis for candidate genesets ################################
bulk.test <- (df.bulk %>% dplyr::select(-genes))
rownames(bulk.test) <- df.bulk$genes

bulk.poi <- GSVA::gsva(as.matrix(bulk.test), go.genes)

# get distance of point from origin
df.mod.path.bulk <- data.frame(t(bulk.poi))
df.mod.path.bulk$line <- stringr::str_extract(rownames(df.mod.path.bulk), "[A-Za-z0-9]*")
df.mod.path.bulk$sample <- rownames(df.mod.path.bulk)

# assign survival data
df.surv <- data.frame(line = c("BT799", "BT935", "BT954", "BT428", "MBT06"),
                      months = c(3, 7.5, 14, 16.5, 35), 
                      status = c("dead", "dead", "dead", "dead", "alive"))

df.mod.path.bulk <- merge(df.mod.path.bulk, df.surv, by = "line")

plt.bulk.gsva.crude <- list()

}

```

```{r SPECIFY GENESET for ANALYSIS HERE}

# get detected CSF protein
df.prot.detect <- read.csv("C:/Users/Owner/Dropbox/PDF Projects - JM/Collaborations/Maleeha/detected_csf_protein_25_300621.csv")
# deg.up <-
ti.set <-  df.ctrl.comp$genes[df.ctrl.comp$avg_log2FC < -1.5]
ag.set <-  df.ctrl.comp$genes[df.ctrl.comp$avg_log2FC > 1.5]

ti.prot <- df.prot.detect$prot[grepl(pattern = paste(ti.set, collapse = "|"), x = df.prot.detect$prot)]
ag.prot <- df.prot.detect$prot[grepl(pattern = paste(ag.set, collapse = "|"), x = df.prot.detect$prot)]
# ag.prot

aset2  <- c()
for (i in 1:length(ag.set)){
 if(sum(grepl(pattern = ag.set[i], x = ag.prot)) > 0){
   aset2 <- c(aset2, ag.set[i])
 }
 # print(a)
}

tset2  <- c()
for (i in 1:length(ti.set)){
 if(sum(grepl(pattern = ti.set[i], x = ti.prot)) > 0){
   tset2 <- c(tset2, ti.set[i])
 }
}

ag.original <- ag.set
ti.original <- ti.set

ti.ag.set <- list(
  ag = ag.set,
  ti = ti.set
)

```





```{r load GBM genesets and public scRNAseq data}

data(geneSets)
verhaak.df <- geneSets[["Verhaak_CancerCell_2010"]]
verhaak.list <- wideDF2namedList(verhaak.df)

gsc.df <- geneSets[["Richards_NatureCancer_2021_sc"]]
gsc.list <- wideDF2namedList(gsc.df)

neftel.df <- geneSets[["GBM_Hs_Neftel2019"]]
neftel.list <- wideDF2namedList(neftel.df)
neftel.list <- neftel.list[!grepl("G", names(neftel.list))]

verhaak.list <- lapply(verhaak.list, toupper)
gsc.list <- lapply(gsc.list, toupper)
neftel.list <- lapply(neftel.list, toupper)

ti.ag.set <- list(
  ag = ag.original,
  ti = ti.original
)

# load in GBM datasets ##########################################################

# load scRNAseq data from Richard et al 2021

load(paste0(dir.root, dir.preprocessed, "R388_M01_NM2_Richards_2021_GSC_080221.Rdata"));

so.richards.gsc <- prepSeurat(so); rm(so)
so.richards.gsc <- downsampleSeurat(so.richards.gsc, 1/3)

# load scRNAseq data from Neftel 2019
load(paste0(dir.root, dir.preprocessed, "R502_M01_NM2_Neftel_2019_10x_lenient_040621.Rdata"));
so.neftel.10x <- prepSeurat(so); rm(so)

```




```{r pathway association helper function}


# get pathway association helper function
 getPathwayAssociation  <- function(object, values, assay = DefaultAssay(object), 
                                    min.pct = 0.5, subsample = 1, cor.method = "spearman", 
                                    species = "Hs", pathway.db = "GO", show.n.pathways = 5, pathway.name.trunc = 30, pos.only = T){
    
    
    message("Preparing expression matrix...")
    expr.mat <- object@assays[[assay]]@data
    if (subsample != 1){
      
      which.ind <- sample(seq(1, ncol(object)), round(ncol(object) * subsample))
      expr.mat <- expr.mat[ ,which.ind]
      values <- values[which.ind]
    }
    
    if (min.pct != 0){
      keep.which <- apply(expr.mat, 1, function(x) mean(x > 0) > min.pct)
      expr.mat <- expr.mat[keep.which, ]
    }
    
    expr.mat.t <- as.matrix(t(expr.mat))
    
    message("Computing correlations...")
    val.cor <- cor(expr.mat.t, values, method = cor.method)
    
    message("Running GSEA...")
    gsea.res <- runGSEA(gene =rownames(val.cor), value = val.cor[ ,1],
                        species = species, db = pathway.db, do.plot = F)
    
    
    # gsea.res <- gsea.res[gsea.res$padj < 0.1, ]
    if (nrow(gsea.res) > 0){

        gsea.res.top <- (gsea.res %>% dplyr::filter(NES > 0) %>% dplyr::arrange(-NES))[1:show.n.pathways, ]
    gsea.res.bottom <- (gsea.res %>% dplyr::filter(NES < 0) %>% dplyr::arrange(NES))[1:show.n.pathways, ]
    
    if (pos.only){
      gsea.res.merge <- gsea.res.top
      threshold.val <- -log10(0.05)
      # x.title <- "log(p)"
    } else {
      gsea.res.merge <- bind_rows(gsea.res.top, gsea.res.bottom)
      # x.title <- "signed(log(p))"
      threshold.val <- c(-log10(0.05), log10(0.05))
    }
    
    
    gsea.res.merge$path.trunc <- stringr::str_trunc(gsea.res.merge$pathway, pathway.name.trunc)
    
    message("Constructing plot...")
    
    plt.gsea <- gsea.res.merge %>%
      ggplot(aes(y = reorder(path.trunc, NES), x = NES, size = -log(pval),fill = abs(NES))) + 
      geom_point(shape = 21) +
      theme_miko(legend = T) + 
      labs(fill = "|NES|", x = "NES", y = "Pathway", fill = "-log(p)") + 
      viridis::scale_fill_viridis()
    
            
    } else {
      plt.gsea <- NULL
    }
      
      return(list(
        cor = val.cor,
        gsea = gsea.res,
        plot = plt.gsea
      ))
    
  }



```

```{r visualize scRNAseq expression of Im signature}

for (i in 1:length(ag.set)){
  try({
     print(exprUMAP(so.query, ag.set[i]))
  }, silent = T)
 
}



```


```{r consensus scoring}

# ag.original
ti.ag.set <- list(
  ag = ag.original,
  ti = ti.original
)


# iteration 1 #################################
which.cor <- "spearman"
sc.res.query1 <- signatureCoherence(object = so.query, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.query), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)
sc.res.neftel1 <- signatureCoherence(object = so.neftel.10x, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.neftel.10x), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)
sc.res.richards1 <- signatureCoherence(object = so.richards.gsc, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.richards.gsc), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)


# sc.res.query$ms.coherent.result$plot.list
sc.res.query1$coherence.plots$ag
sc.res.neftel1$coherence.plots$ag
sc.res.richards1$coherence.plots$ag




# query = sc.res.query$genelist.coherent$ag,
ag.consensus <- lintersect(list(query = sc.res.query1$genelist.coherent$ag,
                           neftel = sc.res.neftel1$genelist.coherent$ag,
                           richards = sc.res.richards1$genelist.coherent$ag))
ti.consensus <- lintersect(list(query = sc.res.query1$genelist.coherent$ti,
                           neftel = sc.res.neftel1$genelist.coherent$ti,
                           richards = sc.res.richards1$genelist.coherent$ti))

# iteration 2 #################################

ti.ag.set <- list(
  ag = ag.consensus,
  ti = ti.consensus
)

sc.res.query <- signatureCoherence(object = so.query, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.query), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)
sc.res.neftel <- signatureCoherence(object = so.neftel.10x, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.neftel.10x), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)
sc.res.richards <- signatureCoherence(object = so.richards.gsc, ms.result = NULL, genelist = ti.ag.set,
                               slot = "data", assay = DefaultAssay(so.richards.gsc), coherence.threshold = 0.1, show.grid = T,
                               coherent.ms = T, cor.method = which.cor)


# sc.res.query$ms.coherent.result$plot.list
sc.res.query$coherence.plots
sc.res.neftel$coherence.plots
sc.res.richards$coherence.plots

# query = sc.res.query$genelist.coherent$ag,
ag.consensus2 <- lintersect(list(query = sc.res.query$genelist.coherent$ag,
                           neftel = sc.res.neftel$genelist.coherent$ag,
                           richards = sc.res.richards$genelist.coherent$ag))
ti.consensus2 <- lintersect(list(query = sc.res.query$genelist.coherent$ti,
                           neftel = sc.res.neftel$genelist.coherent$ti,
                           richards = sc.res.richards$genelist.coherent$ti))

ti.ag.set <- list(
  ag = ag.consensus2,
  ti = ti.consensus2
)

ag.set <- ag.consensus2
ti.set <- ti.consensus2
```

```{r coherence plots, ag signature, fig.width=12, fig.height=10}

plt.coherence <- cowplot::plot_grid(
  sc.res.query1$coherence.plots$ag + theme(legend.position = "bottom") + labs(title = "Current Study"),
  sc.res.neftel1$coherence.plots$ag + theme(legend.position = "bottom")  + labs(title = "Neftel 2019"),
  sc.res.richards1$coherence.plots$ag + theme(legend.position = "bottom")  + labs(title = "Richards 2021"),
    sc.res.query$coherence.plots$ag + theme(legend.position = "bottom") + labs(title = "Current Study"),
  sc.res.neftel$coherence.plots$ag + theme(legend.position = "bottom")  + labs(title = "Neftel 2019"),
  sc.res.richards$coherence.plots$ag + theme(legend.position = "bottom")  + labs(title = "Richards 2021"),
  nrow = 2, labels = "AUTO", rel_heights = c(1.5, 1)
)


plt.coherence


# savePDF("Im_signature_coherence.pdf",plt.coherence, fig.width=12, fig.height=10)
```
