---
title: "barcoding_analysis"
output: 
  flexdashboard::flex_dashboard:
    orientation: rows
    vertical_layout: fill
editor_options: 
  chunk_output_type: inline
---



```{r setup, include=FALSE}

# clear global enviroment
rm(list = ls())

# initiate timer
start.time <- proc.time()

# List of packages to load
packages2load <- c("Seurat", "GSVA", 
                   "plyr", "dplyr", "tidyr", "reshape2", "preprocessCore", "stringr", "RColorBrewer", "ggplot2", "gridExtra", 
                   "DT", "flexdashboard", "ggpmisc", "ggExtra", "grid", "ggrepel", "future", "scMiko", "survival", "survminer", "topGO",
                   "AnnotationDbi", "org.Mm.eg.db", "org.Hs.eg.db", "ddpcr", "tm", "homologene", "plotly", "parallel", "doParallel", "foreach", "quantreg")

# load packages
invisible(lapply(packages2load, library, character.only = TRUE, quietly = T))


```

```{r README}

# LOAD R ENVIROMENT FROM PART 1 PRIOR TO RUNNING THIS SCRIPT

```

```{r FUNCTION score bulk RNA }

# # function to score bulk RNA samples
# # GSVA is run on two query gene sets, and scores are projected onto line of inverse equality (for anti-correlated sets) or equality (for correlated sets)
bulkScoring <- function(df.bulk, so.sig = NULL, gene.set = NULL, set1.sign = 1, set2.sign = -1){
  bulk.test <- (df.bulk %>% dplyr::select(-genes))
  rownames(bulk.test) <- df.bulk$genes

  if (is.null(gene.set) & !is.null(so.sig)){
    deg.gene.sets <- list(
      setA = so.sig$genes[so.sig$direction == 1],
      setB = so.sig$genes[so.sig$direction == -1]
    )
  } else if (!is.null(gene.set) & is.null(so.sig)){

    deg.gene.sets <- gene.set
  }


  deg.gsva.scores <- GSVA::gsva(as.matrix(bulk.test), deg.gene.sets)
  df.bk.gsva.score <- data.frame(y1 = deg.gsva.scores[1,], y2 = deg.gsva.scores[2,], sample = colnames(deg.gsva.scores))
  df.bk.gsva.score$line <- stringr::str_extract(df.bk.gsva.score$sample, "[A-Za-z0-9]*")
  df.bk.gsva.score <- conditionLabels(df.bk.gsva.score)

  sign.multiplier <- 1*(df.bk.gsva.score$y1 > df.bk.gsva.score$y2)
  sign.multiplier[sign.multiplier == 0] <- -1
  sign.multiplier[(1*(df.bk.gsva.score$y1 > df.bk.gsva.score$y2)) == 0] <- set2.sign
  sign.multiplier[(1*(df.bk.gsva.score$y1 > df.bk.gsva.score$y2)) == 1] <- set1.sign

  # proj scores onto point on line
  for (i in 1:nrow(df.bk.gsva.score)){
    if (set1.sign == set2.sign){
      cur.proj <- mosaic::project(x = c(df.bk.gsva.score$y1[i], df.bk.gsva.score$y2[i]), u =c(1,1))
    } else if (set2.sign > set1.sign){
      cur.proj <- mosaic::project(x = c(df.bk.gsva.score$y1[i], df.bk.gsva.score$y2[i]), u =c(1,-1))
    } else if (set1.sign > set2.sign){
      cur.proj <- mosaic::project(x = c(df.bk.gsva.score$y1[i], df.bk.gsva.score$y2[i]), u =c(1,-1))
    }
    #cur.proj <- mosaic::project(x = c(df.bk.gsva.score$y1[i], df.bk.gsva.score$y2[i]), u =c(1,-1))

    df.bk.gsva.score$proj.x[i] <- cur.proj[1]
    df.bk.gsva.score$proj.y[i] <- cur.proj[2]
  }

  # get distance of point from origin
  df.bk.gsva.score$score <- sqrt(df.bk.gsva.score$proj.x^2  + df.bk.gsva.score$proj.y^2)* sign.multiplier

  # assign survival data
  df.surv <- data.frame(line = c("BT799", "BT935", "BT954", "BT428", "MBT06"),
                        months = c(3, 7.5, 14, 16.5, 35),
                        status = c("dead", "dead", "dead", "dead", "alive"))

  df.bk.gsva.score.result <- merge(df.bk.gsva.score, df.surv, by = "line")

  return(df.bk.gsva.score.result)
}
# 

```


```{r FUNCTION plot bulk scores}
# 
plotBulkScores <- function(df.bk.gsva.score, label.1 = NULL, label.2 = NULL, subtitle.label = NULL, which.barcodes = NULL, projection.slope = -1){

  if (is.null(which.barcodes)){
    df.bk.gsva.score2 <- df.bk.gsva.score
  } else {
    df.bk.gsva.score2 <- df.bk.gsva.score %>% dplyr::filter(sample %in% which.barcodes)
  }


plt.deg.gsva.1 <- df.bk.gsva.score2 %>%
  ggplot(aes(x = y1, y = y2, label = sample, color = line)) +
  geom_segment(aes(x = y1, y = y2, xend = proj.x, yend = proj.y)) +
  geom_text(size = 3)+
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0, linetype = "dashed") +
  theme_miko(legend = T) +
  ggthemes::scale_color_tableau() +
  geom_abline(slope = projection.slope) +
  xlab(paste0("GSVA Score (", label.1, ")")) +
  ylab(paste0("GSVA Score (", label.2, ")")) +
  labs(title = "Bulk RNA Scores", subtitle = paste0("R = ", signif(cor(df.bk.gsva.score2$y1, df.bk.gsva.score2$y2),3)))


plt.deg.gsva.2 <- df.bk.gsva.score %>%
  dplyr::filter(sample %in% which.samples) %>%
  ggplot(aes(x = score, y = 0, fill = line)) +
  geom_hline(yintercept = 0, size = 1) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_point( size=3, alpha=0.7, shape=21, stroke=2) +
  theme_miko(legend = T) +
  theme(
    panel.grid.major.x = element_blank(),
    panel.border = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.y = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank()) +
  ggthemes::scale_fill_tableau() +
  labs(title = "Differential Genes", subtitle = subtitle.label) + xlab("Subtype Score")

my.form <- "y~exp(x)"

plt.deg.gsva.3 <- df.bk.gsva.score %>%
  dplyr::filter(sample %in% which.samples) %>%
  ggplot(aes(x = score, y = months)) +
  geom_smooth(method = "lm",  formula = my.form, color = "black", fill = "grey60") +
  geom_point(aes(shape = status), size = 3) +
  geom_text(aes(label = line, color = line), nudge_y = 2, fontface = "bold") +
  theme_miko(legend = T) +
  coord_cartesian(ylim = c(0, max(df.bk.gsva.score$months)*1.1)) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  ggthemes::scale_color_tableau() +
  xlab("Subtype Score") +
  ylab("Survival (Months)") +
  stat_poly_eq(formula = my.form,
               aes(label = paste(..rr.label.., stat(p.value.label),  sep = "~~~")),
               parse = TRUE, rr.digits = 2, p.digits = 2) +
  labs(title = paste0(label.1, "/" ,label.2, " Score vs. Survival"), subtitle = subtitle.label)

# plt.bulk.scores.final <- cowplot::plot_grid(plt.deg.gsva.1, plt.deg.gsva.2, plt.deg.gsva.3, ncol = 1, rel_heights = c(2, 1, 2), align = "hv")


return(list(plt.deg.gsva.1, plt.deg.gsva.2, plt.deg.gsva.3))
}

```


```{r scRNAseq - score cancer subtypes}

# get Cancer cell state genesets
gene.list <- wideDF2namedList( geneSets[["CancerSEA_Hs"]])

# score single cell data #######################################################
csub.mod.score <- matrix(nrow = ncol(so.query), ncol = length(gene.list))
for (i in 1:length(gene.list)){
  
  set.name <- names(gene.list)[i]
  set.list <- gene.list[[i]]
  so.mod <- so.query
  
  set.list <- set.list[(set.list) %in% rownames(so.query)]
  
  so.mod <- AddModuleScore(
    so.mod,
    features = list(set.list),
    nbin = 20,
    ctrl = 100,
    k = FALSE,
    assay = NULL,
    name = set.name,
    seed = 1,
    search = FALSE
  )
  
  csub.mod.score[ ,i] <- so.mod@meta.data[[paste0(set.name, "1")]]

}

rm(so.mod)

colnames(csub.mod.score) <- names(gene.list)
rownames(csub.mod.score) <- colnames(so.query)

# Generate heatmap stratified by sample barcode ################################
bc <- so.query@meta.data[["Barcode"]]
u.bc <- unique(bc)
csub.mod.bc <- matrix(nrow = length(u.bc), ncol = length(gene.list))
for (i in 1:length(u.bc)){
  which.cells <- bc %in% u.bc[i]
  csub.mod.bc[i,] <- colMeans(csub.mod.score[which.cells ,])
}

colnames(csub.mod.bc) <- names(gene.list)
rownames(csub.mod.bc) <- u.bc

# get annotations for heatmap
df.col.ann <- data.frame(sample = rownames(csub.mod.bc))
rownames(df.col.ann) <- rownames(csub.mod.bc)
df.col.ann$line <- NA
df.col.ann$line[grepl("0231", df.col.ann$sample)] <- "MBT06"
df.col.ann$line[grepl("0238", df.col.ann$sample)] <- "BT799"
df.col.ann <- df.col.ann %>% dplyr::select("line")



aka3 = list(line = c(BT799 =  BT799.color, MBT06=MBT06.color))


  scale.lim <- max(abs(csub.mod.bc))

  my.breaks <- seq((-1*scale.lim), scale.lim, by = 0.01)

hm.csub <- pheatmap::pheatmap(t(csub.mod.bc), annotation_col = df.col.ann,
                              breaks = my.breaks,
                                  color = colorRampPalette(rev(brewer.pal(n = 7, name = "RdBu")))(length(my.breaks)), #"BrBG"
                              annotation_colors = aka3, border_color = NA)
# BrBG
plt.hm <- ggplotify::as.ggplot(hm.csub)


print(plt.hm)


# savePDF("final_cancer_state_heatmap_040321.pdf", plt.hm, fig.width = 4, fig.height = 5)

```



```{r scRNAseq - score GBM SUBTYPES (Neftel 2019)}

# get GBM genesets #############################################################
gbm.list <- wideDF2namedList(geneSets[["GBM_Hs_Neftel2019"]])

# score modules ################################################################

module.scores <- matrix(ncol = length(gbm.list), nrow = ncol(so.query))

for (i in 1:length(gbm.list)){
  current.list <- list(g1 = gbm.list[[i]])
  
  so.temp <- AddModuleScore(
    so.query,
    features = current.list,
    pool = NULL,
    nbin = 17,
    ctrl = 100,
    k = FALSE,
    assay = current.assay,
    name = "ModuleScore",
    seed = 1,
    search = FALSE
  )
  
  module.scores[ ,i] <- so.temp@meta.data[["ModuleScore1"]]
}

colnames(module.scores) <- names(gbm.list)
rownames(module.scores) <- rownames(so.query@meta.data)
rm(so.temp)

MES.score <- apply(module.scores[ ,c("MES1", "MES2")], 1, mean)
NPC.score <- apply(module.scores[ ,c("NPC1", "NPC2")], 1, mean)

pooled.scores <- cbind(MES.score, NPC.score)
colnames(pooled.scores) <- c("MES", "NPC")
module.scores <- cbind(module.scores, pooled.scores)

# 2D representation ############################################################

# subset module scores
class.names <- list(
  OPC = "OPC",
  NPC = "NPC",
  AC = "AC",
  MES = "MES"
)


ms <- module.scores[ ,unlist(class.names)]
opc_npc.subset <- c(class.names$OPC, class.names$NPC)
ac_mes.subset <- c(class.names$AC, class.names$MES)

# define y-axis
SC_opc_npc_max <- apply(ms[,opc_npc.subset], 1, function(x) max(x))
SC_ac_mes_max <- apply(ms[,ac_mes.subset], 1, function(x) max(x))

# y-axis value
D.y <- SC_opc_npc_max-SC_ac_mes_max

# define y axis
# x-axis sign class
D.y.sign <- sign(D.y)

# D > 0 class
Dpos.x <- log((abs(ms[,class.names$NPC] - ms[,class.names$OPC]) + 1), base = 2)
Dpos.x.class <- Dpos.x.dif <- ms[,class.names$NPC] - ms[,class.names$OPC]
Dpos.x.class[Dpos.x.dif > 0] <- 1  #NPC
Dpos.x.class[Dpos.x.dif < 0] <- -1 #OPC
Dpos.x <- Dpos.x*Dpos.x.class

# D < 0 class
Dneg.x <- log((abs(ms[,class.names$MES] - ms[,class.names$AC]) + 1), base = 2) 
Dneg.x.class <- Dneg.x.dif <- ms[,class.names$MES] - ms[,class.names$AC]
Dneg.x.class[Dneg.x.dif > 0] <- 1   #MES
Dneg.x.class[Dneg.x.dif < 0] <- -1  #AC
Dneg.x <- Dneg.x*Dneg.x.class


# check if cell cycle scores exist - if not, score. 
df.state <- data.frame(D.y, 
                       Dpos.x, 
                       Dneg.x, 
                       D.y.sign, 
                       barcode = so.query@meta.data[["Barcode"]],
                       cluster = so.query@meta.data[["seurat_clusters"]],
                       cell.cycle = so.query@meta.data[["Phase"]]
                       )

df.state$y <- df.state$D.y
df.state$x <- df.state$Dpos.x
df.state$x[df.state$y < 0] <- df.state$Dneg.x[df.state$y < 0]
scale.max <- max(c(abs(df.state$x), abs(df.state$y))) * 1.1

# Generate PLOT
df.labels <- data.frame(x = c(scale.max*0.9,-scale.max*0.9,-scale.max*0.9,scale.max*0.9),
                        y = c(scale.max,scale.max,-scale.max,-scale.max),
                        label = c("NPC", "OPC", "AC", "MES"))

color.pal <- "grey" # lightgrey

# line class
df.state$line <- NA
df.state$line[grepl("0231", df.state$barcode)] <- "MBT06"
df.state$line[grepl("0238", df.state$barcode)] <- "BT799"

# GBM classification
df.state$subtype <- NA
df.state$subtype[df.state$x < 0 & df.state$y < 0] <- "AC"
df.state$subtype[df.state$x < 0 & df.state$y > 0] <- "OPC"
df.state$subtype[df.state$x > 0 & df.state$y < 0] <- "MES"
df.state$subtype[df.state$x > 0 & df.state$y > 0] <- "NPC"

df.state.line <- df.state %>%
  dplyr::group_by(line, barcode) %>%
  dplyr::summarize(x.score = mean(x), 
                   y.score = mean(y))

df.state.sub <- df.state %>%
  dplyr::group_by(line, barcode, subtype) %>%
  dplyr::summarize(count.x = length(x))



plt.metascores <- df.state %>%
  ggplot(aes(x, y)) + 
  xlab("Relative meta-module score\n[log(|SC1-SC2|+1)]") + 
  ylab("Relative meta-module score\n[log(|SC1-SC2|+1)]") + 
  labs(title = "GBM Subtypes", subtitle = "2D Representation (Neftel 2019)") + 
  theme_miko(legend = T)  +
  theme(panel.border = element_rect(colour = color.pal, fill=NA, size=4)) +
  geom_hline(yintercept=0, color = color.pal, size=1.5) +
  geom_vline(xintercept=0, color = color.pal, size=1.5) +
  geom_point(aes(color = barcode), size = 0.9, alpha = 0.8) +
  geom_point(data = df.state.line, aes(x = x.score, y = y.score, fill = barcode), size = 4, pch=21) + 
  geom_label(data = df.labels, aes(x = x, y = y, label = label), fill = color.pal, color="white") + 
  ggthemes::scale_colour_tableau() + 
   ggthemes::scale_fill_tableau() + 
  facet_wrap(~line)

plt.metascores

# chi square test ##############################################################
runXsq <- function(df.data, subgroup = NULL){
  
  df.data.sub <- df.data %>% dplyr::filter(line == subgroup)
  df.data.sub <- df.data.sub %>% group_by(barcode) %>% mutate(freq.x = count.x/sum(count.x))
  df.data.sub <- df.data.sub %>% dplyr::select(c("line", "subtype", "freq.x", "barcode"))
  bt.bar.u <- unique(df.data.sub$barcode)
  df.data.sub.wide <- as.data.frame(pivot_wider(df.data.sub, values_from = freq.x, names_from = barcode))
  df.data.sub.wide[ ,bt.bar.u[1]][is.na(df.data.sub.wide[ ,bt.bar.u[1]])] <- 0
  df.data.sub.wide[ ,bt.bar.u[2]][is.na(df.data.sub.wide[ ,bt.bar.u[2]])] <- 0
  
  Xsq.sub <- chisq.test(x = df.data.sub.wide[ ,bt.bar.u[2]], p =df.data.sub.wide[ ,bt.bar.u[1]], correct = TRUE,
                        rescale.p = FALSE, simulate.p.value = F, B = 2000)
  
  return(Xsq.sub)
  
}

Xsq.BT799 <- runXsq(df.state.sub, subgroup = "BT799")
Xsq.MBT06 <- runXsq(df.state.sub, subgroup = "MBT06")

# generate compositional barplots ##############################################
plt.bar1 <- df.state.sub %>%
  dplyr::filter(line == "BT799") %>%
  ggplot(aes(x = barcode, y = count.x, fill = subtype)) + 
  geom_bar(stat = "identity", position = "fill", alpha = 0.8) + 
  theme_miko() + 
  ggthemes::scale_fill_economist() + 
  xlab("Sample") + 
  ylab("Relative Frequency") + 
  labs(title = "Subtypes", subtitle = paste0("BT799\np=", signif(Xsq.BT799[["p.value"]], 3))) 

plt.bar2 <- df.state.sub %>%
  dplyr::filter(line == "MBT06") %>%
  ggplot(aes(x = barcode, y = count.x, fill = subtype)) + 
  geom_bar(stat = "identity", position = "fill", alpha = 0.8) + 
  theme_miko(legend = T) + 
  ggthemes::scale_fill_economist() + 
  xlab("Sample") + 
  ylab("Relative Frequency")  + 
  labs(subtitle = paste0("MBT06\np=", signif(Xsq.MBT06[["p.value"]]))) 

plt.bar.merge <- cowplot::plot_grid(plt.bar1, plt.bar2, align = "hv", ncol = 2, rel_widths = c(1.55,2))

print(plt.bar.merge)

```
